src/firstboot/firstboot.c

```diff
@@ -774,93 +777,68 @@ static int write_root_shadow(const char *shadow_path, const char *hashed_passwor
 static int process_root_args(void) {
         _cleanup_close_ int lock = -1;
         struct crypt_data cd = {};
-        const char *hashed_password;
-        const char *etc_shadow;
+        const char *password, *hashed_password;
+        const char *etc_passwd, *etc_shadow;
         int r;
 
+        etc_passwd = prefix_roota(arg_root, "/etc/passwd");
         etc_shadow = prefix_roota(arg_root, "/etc/shadow");
-        if (laccess(etc_shadow, F_OK) >= 0 && !arg_force)
+
+        /* Reduce the number of possible scenarios by always making sure we have permission to modify both
+         * /etc/passwd and /etc/shadow. */
+        if ((laccess(etc_passwd, F_OK) >= 0 || laccess(etc_shadow, F_OK) >= 0) && !arg_force)
                 return 0;
 
-        (void) mkdir_parents(etc_shadow, 0755);
+        (void) mkdir_parents(etc_passwd, 0755);
 
         lock = take_etc_passwd_lock(arg_root);
         if (lock < 0)
-                return log_error_errno(lock, "Failed to take a lock: %m");
+                return log_error_errno(lock, "Failed to take a lock on %s: %m", etc_passwd);
 
         if (arg_copy_root_shell && arg_root) {
                 struct passwd *p;
 
                 errno = 0;
                 p = getpwnam("root");
-                if (!p && errno != ENOENT) {
-                        if (!errno)
-                                errno = EIO;
-
+                if (!p)
                         return log_error_errno(errno, "Failed to find passwd entry for root: %m");
-                }
 
-                if (p) {
-                        r = free_and_strdup(&arg_root_shell, p->pw_shell);
-                        if (r < 0)
-                                return log_oom();
-                }
+                r = free_and_strdup(&arg_root_shell, p->pw_shell);
+                if (r < 0)
+                        return log_oom();
         }
 
         r = prompt_root_shell();
         if (r < 0)
                 return r;
 
-        if (arg_delete_root_password || arg_root_shell) {
-                const char *etc_passwd;
-
-                /* Mixing alloca() and other stuff that touches the stack in one expression is not portable. */
-                etc_passwd = prefix_roota(arg_root, "/etc/passwd");
-
-                r = write_root_passwd(etc_passwd, arg_delete_root_password ? "" : NULL, arg_root_shell);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to write %s: %m", etc_passwd);
-
-                log_info("%s written", etc_passwd);
-
-                return 0;
-        }
-
         if (arg_copy_root_password && arg_root) {
                 struct spwd *p;
 
-                errno = 0;
```

Comment:
this should stay. errno handling is notoriously unreliable with password handling function (NSS modules) and documentation vague

Reply:
I should have read the docs first.